kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/records/RecordStream.java

```diff
@@ -0,0 +1,229 @@
+                    result.add(newState);
+                }
+            }
+        }
+        return new RecordStream<>(records, PairType.PER_RECORD, result);
+    }
+
+    public void forEachRecord(RecordConsumer<T> mapper) {
+        int i = 0;
+        for (var batch : records.batches()) {
+            if (!batch.isControlBatch()) {
+                for (var record : batch) {
+                    T existingState = existingState(i++);
+                    mapper.accept(batch, record, existingState);
+                }
+            }
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private T existingState(int recordIndex) {
+        return switch (pairType) {
+            case SINGLE -> (T) pairedWith;
+            case INDEX -> (T) Integer.valueOf(recordIndex);
+            case PER_RECORD -> ((List<T>) pairedWith).get(recordIndex);
+        };
+    }
+
+    /**
+     * Map each of the records in this stream to some new state and return the set of those mapped states.
+     * This iterates the batches in the source {@link MemoryRecords} and so will result in
+     * batch decompression.
+     * @param mapper The mapper function
+     * @return The set
+     * @param <S> The type of state
+     */
+    public <S> Set<S> toSet(RecordMapper<T, S> mapper) {
+        var result = new HashSet<S>();
+        int i = 0;
+        for (var batch : records.batches()) {
+            if (!batch.isControlBatch()) {
+                for (var record : batch) {
+                    T existingState = existingState(i++);
+                    result.add(mapper.apply(batch, record, existingState));
+                }
+            }
+        }
+        return result;
+    }
+
+    public <S> List<S> toList(RecordMapper<T, S> mapper) {
+        var result = new ArrayList<S>();
+        int i = 0;
+        for (var batch : records.batches()) {
+            for (var record : batch) {
+                if (!batch.isControlBatch()) {
+                    T existingState = existingState(i++);
+                    result.add(mapper.apply(batch, record, existingState));
+                }
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Applies a {@link RecordTransform} to the records in this stream,
+     * returning the mapped recurds in a {@link MemoryRecords}.
+     * This method will preserve empty batches and control batches.
+     * This iterates the batches in the source {@link MemoryRecords} and so will result in
+     * batch decompression.
+     *
+     * @param transform The record transform
+     * @return The mapped records
+     */
+    public MemoryRecords toMemoryRecords(@NonNull ByteBufferOutputStream buffer,
+                                         @NonNull RecordTransform<T> transform) {
+        BatchAwareMemoryRecordsBuilder builder = new BatchAwareMemoryRecordsBuilder(buffer);
+        int indexInStream = 0;
+        for (var batch : records.batches()) {
+            if (batch.isControlBatch()) {
```

Comment:
Have added unit tests for control and empty cases in https://github.com/tombentley/kproxy/pull/14